frontend/src/main/scala/bloop/engine/tasks/CompileTasks.scala

```diff
@@ -0,0 +1,117 @@
+import sbt.internal.inc.{ConcreteAnalysisContents, FileAnalysisStore}
+
+object CompileTasks {
+  import bloop.engine.State
+  def compile(state: State, project: Project, reporterConfig: ReporterConfig): State = {
+    import State.compilerCache
+    import state.logger
+    def toInputs(project: Project, config: ReporterConfig, result: PreviousResult) = {
+      val instance = project.scalaInstance
+      val sourceDirs = project.sourceDirectories
+      val classpath = project.classpath
+      val classesDir = project.classesDir
+      val target = project.tmp
+      val scalacOptions = project.scalacOptions
+      val javacOptions = project.javacOptions
+      val reporter = new Reporter(logger, project.baseDirectory.syntax, identity, config)
+      // FORMAT: OFF
+      CompileInputs(instance, compilerCache, sourceDirs, classpath, classesDir, target, scalacOptions, javacOptions, result, reporter, logger)
+      // FORMAT: ON
+    }
+
+    val results = new java.util.concurrent.ConcurrentHashMap[Project, PreviousResult]()
+    def runCompilation(project: Project): Seq[Project] = {
+      val previousResult = state.results
+        .getResult(project)
+        .getOrElse(sys.error("Results cache was not initialized"))
+      val inputs = toInputs(project, reporterConfig, previousResult)
+      val result = Compiler.compile(inputs)
+      results.put(project, result)
+      List(project)
+    }
+
+    def compileTask(project: Project) =
+      new Task((_: Seq[Project]) => runCompilation(project), () => ())(Mergeable.SeqMergeable)
+
+    import bloop.engine.{Leaf, Parent}
+    val visitedTasks = scala.collection.mutable.HashMap[Dag[Project], Task[Seq[Project]]]()
+    def constructTaskGraph(dag: Dag[Project]): Task[Seq[Project]] = {
+      def createTask: Task[Seq[Project]] = {
+        dag match {
+          case Leaf(project) => compileTask(project)
+          case Parent(project, children) =>
+            val childrenCompilations = children.map(constructTaskGraph)
+            val parentCompilation = compileTask(project)
+            childrenCompilations.foldLeft(parentCompilation) {
+              case (task, childrenTask) => task.dependsOn(childrenTask); task
+            }
+        }
+      }
+
+      visitedTasks.get(dag) match {
+        case Some(task) => task
+        case None =>
+          val task = createTask
+          visitedTasks.put(dag, task)
+          task
+      }
+    }
+
+    def updateState(state: State): State = {
+      import scala.collection.JavaConverters._
+      val updatedResults = results.asScala.iterator.foldLeft(state.results) {
+        case (results, (project, result)) => results.updateCache(project, result)
+      }
+      state.copy(results = updatedResults)
+    }
+
+    val taskGraph = constructTaskGraph(state.build.getDagFor(project))
+    Await.result(taskGraph.run()(state.executionContext), Duration.Inf) match {
+      case _: Task.Success[_] => updateState(state)
```

Comment:
Why are we ignoring the result of the task and putting stuff in `result`? Why is that better than simply getting the result out of the task?

Reply:
You're right the other one is better, will change